src/internal/log/loggers.go

```diff
@@ -57,6 +64,41 @@ func SetGRPCLogLevel(l zapcore.Level) {
 	grpcLevel.SetLevel(l)
 }
 
+type atomicLeveler interface {
+	Level() zapcore.Level
+	SetLevel(zapcore.Level)
+}
+
+// revertLogLevel sets up a log level change that is reverted to an original level after the
+// duration, logging the revert with the text of msg, ensuring that even with multiple concurrent
+// calls, the most recent call cancels the actions of prior calls.
+func revertLogLevel(tPtr *atomic.Pointer[time.Timer], levelVar atomicLeveler, originalLevel, newLevel zapcore.Level, d time.Duration, msg string) {
+	wasSet := make(chan struct{})
+	oldTimer := tPtr.Swap(time.AfterFunc(d, func() {
+		<-wasSet
+		cur := levelVar.Level()
+		levelVar.SetLevel(originalLevel)
+		zap.L().Info(msg, zap.Stringer("from", cur), zap.Stringer("to", originalLevel))
+	}))
+	if oldTimer != nil {
+		oldTimer.Stop()
```

Comment:
Do we have to drain a channel here?

Reply:
Nope.  time.NewTimer requires that, but the channel is not part of a time.Timer itself, it's something NewTimer adds.